actix-server/src/worker.rs

```diff
@@ -38,35 +38,114 @@ pub(crate) struct Stop {
 #[derive(Debug)]
 pub(crate) struct Conn {
     pub io: MioStream,
-    pub token: Token,
+    pub token: usize,
 }
 
 fn handle_pair(
     idx: usize,
     tx1: UnboundedSender<Conn>,
     tx2: UnboundedSender<Stop>,
-    avail: WorkerAvailability,
+    counter: Counter,
 ) -> (WorkerHandleAccept, WorkerHandleServer) {
-    let accept = WorkerHandleAccept { tx: tx1, avail };
+    let accept = WorkerHandleAccept {
+        idx,
+        tx: tx1,
+        counter,
+    };
 
     let server = WorkerHandleServer { idx, tx: tx2 };
 
     (accept, server)
 }
 
+#[derive(Clone)]
+pub(crate) struct Counter {
+    counter: Arc<AtomicUsize>,
+    limit: usize,
+}
+
+impl Counter {
+    pub(crate) fn new(limit: usize) -> Self {
+        Self {
+            counter: Arc::new(AtomicUsize::new(1)),
+            limit,
+        }
+    }
+
+    /// Increment counter it by 1 and return false when hitting limit
+    #[inline(always)]
+    pub(crate) fn incr(&self) -> bool {
+        self.counter.fetch_add(1, Ordering::Relaxed) != self.limit
```

Comment:
condition should be on >= limit otherwise there's a race condition on the increment causing missed signalling of non-readiness.


would also suggest creating a simple loom test for it; my attempts seem to suggest that relaxed is okay here but it feels wrong

Reply:
There is no race as Accept is the only thread that increment it. The order does not really matter also because as soon as counter equals to limit a worker is marked as un avail.

We only have two thread operate on certain counter and there is no 3rd thread get in the way to inc or dec the counter.

Reply:
This makes sense. Would appreciate it being written up in the code so others can follow the logic later.

Reply:
Added some doc comment on the Counter type. Let me know if it's clear.